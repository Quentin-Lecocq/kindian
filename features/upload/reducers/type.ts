import { Book } from '../type';

export type FormState = {
  error: string | null;
  downloadUrl: string[];
  isUploading: boolean;
  books: Book[];
  fileContent: string | null;
};

export type FormAction =
  | { type: 'SET_UPLOADING'; payload: boolean }
  | { type: 'SET_ERROR'; payload: string }
  | { type: 'SET_DOWNLOAD_URL'; payload: string[] }
  | { type: 'SET_BOOKS'; payload: Book[] }
  | { type: 'SET_FILE_CONTENT'; payload: string };
