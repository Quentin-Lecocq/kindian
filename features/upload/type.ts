export interface Book {
  title: string;
  highlights: Highlight[];
  selected?: boolean;
}

export interface Highlight {
  info: string;
  quote: string;
}
