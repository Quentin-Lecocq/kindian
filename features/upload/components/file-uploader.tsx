'use client';

import { Button } from '@/components/ui/button';
import { cn } from '@/lib/utils';
import { DownloadIcon, FileDownIcon, Upload } from 'lucide-react';
import Dropzone from 'react-dropzone';
import ErrorDisplay from './error-display';
import TableBooks from './table-books';
import useFileUploaderPresenter from './use-file-uploader.presenter';

const FileUploader = () => {
  const {
    books,
    error,
    handleFileChange,
    handleSelectBook,
    handleToggleSelectAllBooks,
    handleResetUploader,
    handleExportAndDownload,
  } = useFileUploaderPresenter();

  return (
    <div className="flex flex-col gap-12 p-10 items-center justify-center h-screen">
      <div className="w-full mt-10 flex h-60 items-center justify-center">
        {!books.length ? (
          <Dropzone
            accept={{
              'text/plain': ['.txt'],
            }}
            onDrop={handleFileChange}
            maxFiles={1}
          >
            {({ getRootProps, getInputProps, open }) => (
              <div
                {...getRootProps()}
                className={cn(
                  'group relative grid h-60 w-full cursor-pointer place-items-center rounded-lg border-2 border-dashed border-muted-foreground/25 px-5 py-2.5 text-center transition hover:bg-muted/25',
                  'ring-offset-background focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2'
                )}
              >
                <input {...getInputProps()} />
                <div className="flex flex-col gap-4">
                  <div className="flex flex-col items-center justify-center gap-4 sm:px-5">
                    <div className="rounded-full border border-dashed p-3">
                      <Upload
                        className="size-7 text-muted-foreground"
                        aria-hidden="true"
                      />
                    </div>
                    <div className="flex flex-col gap-px">
                      <p className="font-medium text-muted-foreground">
                        Drag {`'n'`} drop files here, or click to select files
                      </p>
                      <p className="text-sm text-muted-foreground/70">
                        You can upload a txt file directly from your kindle
                      </p>
                    </div>
                  </div>
                  <Button
                    className="w-full bg-emerald-400"
                    onClick={() => open()}
                  >
                    Upload File
                  </Button>
                </div>
              </div>
            )}
          </Dropzone>
        ) : (
          <div className="w-96 flex flex-col gap-4">
            <h4 className="text-2xl font-bold">Actions available</h4>
            <p className="text-sm text-muted-foreground">
              Books imported successfully. Select an export option below.
            </p>
            <div className="flex gap-4">
              <Button
                className="w-full bg-emerald-400"
                onClick={() => handleExportAndDownload(false)}
              >
                <FileDownIcon />
                Export selected books
              </Button>
              <Button
                className="w-full"
                onClick={async () => handleExportAndDownload(true)}
              >
                <DownloadIcon />
                Export all books
              </Button>
            </div>
            <Button
              onClick={handleResetUploader}
              variant="link"
              className="p-0 self-start text-muted-foreground"
            >
              reset uploader
            </Button>
          </div>
        )}
      </div>

      {error && <ErrorDisplay error={error} />}

      <div className="border-t flex-1 w-full flex">
        <div className="w-full">
          {books.length > 0 && (
            <TableBooks
              books={books}
              onSelect={handleSelectBook}
              onToggleSelectAll={handleToggleSelectAllBooks}
            />
          )}
        </div>
      </div>
    </div>
  );
};

export default FileUploader;
