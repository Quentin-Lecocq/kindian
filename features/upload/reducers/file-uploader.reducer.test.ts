import { Book } from '../type';
import { fileUploaderReducer } from './file-uploader.reducer';
import { FileUploaderAction, FileUploaderState } from './type';

const initialState: FileUploaderState = {
  error: null,
  isUploading: false,
  books: [],
  fileContent: null,
};

describe('fileUploaderReducer', () => {
  it('should handle SET_UPLOADING', () => {
    const action: FileUploaderAction = { type: 'SET_UPLOADING', payload: true };
    const newState = fileUploaderReducer(initialState, action);
    expect(newState.isUploading).toBe(true);
  });

  it('should handle SET_ERROR', () => {
    const action: FileUploaderAction = { type: 'SET_ERROR', payload: 'error' };
    const newState = fileUploaderReducer(initialState, action);
    expect(newState.error).toBe('error');
  });

  it('should handle SET_BOOKS', () => {
    const action: FileUploaderAction = { type: 'SET_BOOKS', payload: [] };
    const newState = fileUploaderReducer(initialState, action);
    expect(newState.books).toEqual([]);
  });

  it('should add books to the state', () => {
    const book: Book = {
      id: '1',
      title: 'Test Book',
      author: 'author',
      highlights: [],
    };

    const action: FileUploaderAction = { type: 'SET_BOOKS', payload: [book] };
    const newState = fileUploaderReducer(initialState, action);
    expect(newState).toMatchSnapshot();
  });

  it('should handle SET_FILE_CONTENT', () => {
    const action: FileUploaderAction = {
      type: 'SET_FILE_CONTENT',
      payload: 'fileContent',
    };
    const newState = fileUploaderReducer(initialState, action);
    expect(newState.fileContent).toBe('fileContent');
  });

  it('should throw an error for an unhandled action type', () => {
    const action = {
      type: 'UNKNOWN_ACTION',
      payload: 'payload',
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
    } as any;

    expect(() => fileUploaderReducer(initialState, action)).toThrow(
      'Unhandled action type'
    );
  });

  it('should not mutate the initial state', () => {
    const action: FileUploaderAction = { type: 'SET_UPLOADING', payload: true };
    const newState = fileUploaderReducer(initialState, action);
    expect(newState).not.toBe(initialState);
    expect(initialState.isUploading).toBe(false);
  });
});
