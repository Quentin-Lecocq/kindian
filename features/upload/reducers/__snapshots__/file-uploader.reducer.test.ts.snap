// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`fileUploaderReducer > should add books to the state 1`] = `
{
  "books": [
    {
      "author": "author",
      "highlights": [],
      "id": "1",
      "title": "Test Book",
    },
  ],
  "error": null,
  "fileContent": null,
  "isUploading": false,
}
`;
